rfcs/2021-05-17-7469-load-balancing-aggregators.md

```diff
@@ -0,0 +1,99 @@
+* Using a dedicated reverse proxy to load balance requests for Vector aggregators should support the largest spread of `sources` with the smallest amount of engineering effort.
+* A solution outside of Vector itself ensures that users can reliably adopt Vector as an aggregator without replacing their existing infrastructure.
+* Most organizations are likely familiar with operating _some_ class of reverse proxy.
+* Dedicated reverse proxy can be specialized and optimized for its task, and the same can be said for Vector itself.
+
+## Prior Art
+
+* [Logstash: Scaling TCP, UDP, and HTTP](https://www.elastic.co/guide/en/logstash/current/deploying-and-scaling.html#_tcp_udp_and_http_protocols)
+* [Fluentd: Aggregator behind Network Load Balancer](https://aws.amazon.com/blogs/compute/building-a-scalable-log-solution-aggregator-with-aws-fargate-fluentd-and-amazon-kinesis-data-firehose/)
+
+## Drawbacks
+
+* The team will need to maintain a configuration for a third-party application, as well as ensuring the application is kept up-to-date and free of any reported vulnerabilities.
+* We will also need to add the reverse proxy to new or existing integration tests to ensure there are no regressions with our provided configuration and proxy version.
+* Our deployment will be more complex and require an additional application for end users. This can create more misdirection while debugging and additional operational burden.
+
+## Alternatives
+
+### Do Nothing
+
+The Vector aggregator can currently function as a single instance and be scaled vertically rather than horizontally. While this reduces complexity, it causes Vector to be a single point of failure and introduces an upper limit for throughput.
+
+### Only Client-side Load Balancing
+
+The library powering the v2 Vector `sink`/`source` does provide the capabilities to do client-side load balancing, however that just covers a single `sink` to `source` pairing. For certain clients like Beats and Logstash we could implement an Elasticsearch compatible API and allow those clients to use their native load balancing and integrations, this would generally be per source and not available for all sources.
+
+### Require a Service Mesh
+
+Users already leveraging a service mesh could offload the load balancing to the mesh, however, requiring a service mesh to run and scale Vector aggregators horizontally is a large barrier to adoption.
+
+### Distributed hashring
+
+Project like Thanos and Loki have used hashrings to enable multi-tenancy, we could likely do something similar to ensure events are forwarded to the correct aggregator. I don't think anyone wants to turn Vector into a distributed system though.
+
+## Outstanding Questions
+
+* [ ] Which reverse proxy to use? HAProxy, NGINX, Envoy, Traefik, etc. It should be widely used, battle-tested, support most/all protocols Vector uses, and preferably well understood by mutliple members of our team.
```

Comment:
We should pick one and continue iterating until we feel the quality is sufficient, but I would like to support multiple proxies in the future, this way we can slot in easier to existing setups. This can be an option in our Helm charts, right?

Reply:
Yes, we could include any number of proxies as sub-charts and enable/disable as needed - it'll just expand our testing area for the charts.

Reply:
Agreed, I think we could start with one, but let users configure others. For example, if they are using a cloud provider's managed k8s service they may prefer to use the default load balancer for that cloud provider.